Idrall/Resolve.idr

```diff
@@ -3,75 +3,137 @@ module Idrall.Resolve
 import Idrall.Error
 import Idrall.Expr
 import Idrall.IOEither
+import Idrall.Parser
+import Idrall.Path
+
+liftEither : Either e a -> IOEither e a
+liftEither = MkIOEither . pure
+
+mapErr : (e -> e') -> IOEither e a -> IOEither e' a
+mapErr f (MkIOEither x) = MkIOEither (do
+  x' <- x
+  (case x' of
+        (Left l) => pure (Left (f l))
+        (Right r) => pure (Right r)))
+
+parseErrorHandler : String -> Error
+parseErrorHandler x = ErrorMessage (x)
+
+fileErrorHandler : String -> FileError -> Error
+fileErrorHandler x y = ReadFileError (show y ++ " " ++ x)
+
+readFile' : String -> IOEither Error String
+readFile' x =
+  let c = MkIOEither (readFile x)
+      contents = mapErr (fileErrorHandler x) c in
+  contents
+
+nextCurrentPath : (current : Maybe Path) -> (next : Path) -> Path
+nextCurrentPath (Just (Home xs)) (Relative ys) = Home (xs ++ ys)
+nextCurrentPath (Just (Absolute xs)) (Relative ys) = Absolute (xs ++ ys)
+nextCurrentPath (Just (Relative xs)) (Relative ys) = Relative (xs ++ ys)
+nextCurrentPath _ p = p
+
+combinePaths : Maybe FilePath -> FilePath -> FilePath
+combinePaths Nothing y = y
+combinePaths (Just (MkFilePath pathX fileNameX)) (MkFilePath pathY fileNameY) =
+  let nextPath = nextCurrentPath (Just pathX) pathY
+  in
+  MkFilePath nextPath fileNameY
+
+canonicalFilePath : FilePath -> String -- TODO finish properly
+canonicalFilePath x = filePathForIO x
+
+alreadyImported : List FilePath -> FilePath -> Either Error () -- TODO check is correct
+alreadyImported xs x = case elem x xs of
+                            False => Right ()
+                            True => Left (CyclicImportError ((show x) ++ " in " ++ (show xs)))
 
 mutual
+  resolveLocalFile : (history : List FilePath) -> (current : Maybe FilePath) -> (next : FilePath) -> IOEither Error (Expr Void)
+  resolveLocalFile h current next =
+    let combinedFilePaths = combinePaths current next in
+        go combinedFilePaths
+    where
+    go : FilePath -> IOEither Error (Expr Void)
+    go p = do
+      liftEither (alreadyImported h (normaliseFilePath p))
+      str <- readFile' (canonicalFilePath p)
+      expr <- mapErr (parseErrorHandler) (liftEither (parseExpr str))
+      resolve (normaliseFilePath p :: h) (Just p) expr
+
   export
-  resolve : Expr ImportStatement -> IOEither Error (Expr Void)
-  resolve e@(EVar x) = pure e
-  resolve e@(EConst x) = pure e
-  resolve (EPi x y z) = do
-    y' <- resolve y
-    z' <- resolve z
+  resolve : (history : List FilePath) -> Maybe FilePath -> Expr ImportStatement -> IOEither Error (Expr Void)
```

Comment:
Still thinking that this looks a lot like `traverse` :P

Reply:
Future me will make that happen somehow 😅